templates/servicemonitor.yaml

```diff
@@ -0,0 +1,97 @@
+    targetPort: {{ .Values.metrics.serviceMonitor.targetPort }}
+    {{- else -}}
+    port: {{ .Values.metrics.servicePortName | quote }}
+    {{- end }}
+    {{- if .Values.metrics.serviceMonitor.path }}
+    path: {{ .Values.metrics.serviceMonitor.path | quote }}
+    {{- end }}
+    {{- if .Values.metrics.serviceMonitor.scheme }}
+    scheme: {{ .Values.metrics.serviceMonitor.scheme | quote }}
+    {{- end }}
+    {{- if .Values.metrics.serviceMonitor.params }}
+    params:
+      {{- toYaml .Values.metrics.serviceMonitor.params | nindent 6 }}
+    {{- end }}
+    {{- if .Values.metrics.serviceMonitor.interval }}
+    interval: {{ .Values.metrics.serviceMonitor.interval | quote }}
+    {{- end }}
+    {{- if .Values.metrics.serviceMonitor.scrapeTimeout }}
+    scrapeTimeout: {{ .Values.metrics.serviceMonitor.scrapeTimeout | quote }}
+    {{- end }}
+    {{- if .Values.metrics.serviceMonitor.tlsConfig }}
+    tlsConfig:
+      {{- toYaml .Values.metrics.serviceMonitor.tlsConfig | nindent 6 }}
+    {{- end }}
+    {{- if .Values.metrics.serviceMonitor.bearerTokenFile }}
+    bearerTokenFile: {{ .Values.metrics.serviceMonitor.bearerTokenFile | quote }}
+    {{- end }}
+    {{- if .Values.metrics.serviceMonitor.bearerTokenSecret }}
+    bearerTokenSecret:
+      {{- toYaml .Values.metrics.serviceMonitor.bearerTokenSecret | nindent 6 }}
+    {{- end }}
+    {{- if .Values.metrics.serviceMonitor.honorLabels }}
+    honorLabels: {{ .Values.metrics.serviceMonitor.honorLabels }}
+    {{- end }}
+    {{- if .Values.metrics.serviceMonitor.honorTimestamps }}
+    honorTimestamps: {{ .Values.metrics.serviceMonitor.honorTimestamps }}
+    {{- end }}
+    {{- if .Values.metrics.serviceMonitor.basicAuth }}
+    basicAuth:
+      {{- toYaml .Values.metrics.serviceMonitor.basicAuth | nindent 6 }}
+    {{- end }}
+    {{- if .Values.metrics.serviceMonitor.metricRelabelings }}
+    metricRelabelings:
+      {{- toYaml .Values.metrics.serviceMonitor.metricRelabelings | nindent 6 }}
+    {{- end }}
+    {{- if .Values.metrics.serviceMonitor.relabelings }}
+    relabelings:
+      {{- toYaml .Values.metrics.serviceMonitor.relabelings | nindent 6 }}
+    {{- end }}
+    {{- if .Values.metrics.serviceMonitor.proxyUrl -}}
+    proxyUrl: {{ .Values.metrics.serviceMonitor.proxyUrl }}
+    {{- end }}
+  selector:
+    matchLabels:
+      app: {{ include "docker-registry.name" . | quote }}
+      chart: {{ .Chart.Name }}-{{ .Chart.Version | replace "+" "_" }}
+      release: {{ .Release.Name | quote }}
+      heritage: {{ .Release.Service | quote }}
+  namespaceSelector:
+    matchNames:
+      - {{ .Release.Namespace | quote }}
+  {{- if .Values.metrics.serviceMonitor.sampleLimit }}
+  sampleLimit: {{ .Values.metrics.serviceMonitor.sampleLimit }}
+  {{- end }}
+  {{- if .Values.metrics.serviceMonitor.targetLimit }}
+  targetLimit: {{ .Values.metrics.serviceMonitor.targetLimit }}
+  {{- end }}
+{{- end }}
+{{- end }}
```

Comment:
This template feels like overkill to me. If every single value is configurable, then it offers little benefit over a user just writing a separate manifest and deploying that. Instead, I'd propose just writing the bare minimum necessary for a functional service monitor in this template, and reduce its configurability to the bare essentials. For fine-grained configurability, the user can disable this chart from deploying a service monitor and the user can deploy their own alongside this chart rather than through it.